### ISSUE TITLE
Issue after immediate lock releasing for partitions upon partitioned table deletion

### ISSUE THREAD

**Issue by maksm90** [greenplum-db/gpdb#12962]
Commit https://github.com/greenplum-db/gpdb/commit/fa287d01b8b8f989b85d1ad69a2f42da6d762b4a had introduced the logic restricting unbounded lock acquiring for partitions upon different exclusive operations on parent table. For example, deletion of partitioned table [incurs](https://github.com/greenplum-db/gpdb/blob/6.19.0/src/backend/catalog/dependency.c#L1346-L1358) acquiring and immediately releasing of AccessExclusive locks on partitions leaving the same kind lock on partitioned table until transaction completion. This might lead to unexpected behavior when concurrent transaction performs access to some partition explicitly bypassing acquiring some lock on parent table. For example, we have gotten segfault upon concurrent deletion of partitioned table and calling `pg_total_relation_size()` on some partition. This failure might be reproduced as follows:
```sql
# Prepare setup: create appendonly partitioned table with two partitions
0: create table ptest(i int) with(appendonly=true) partition by range (i) (start (0) end (2) exclusive every (1));

# The first session removes partitioned table inside transaction block
1: begin;
1: drop table ptest;

# The second session executes pg_total_relation_size() up to accessing to pg_aoseg relation for some partition on some segment node
(gdb) break dbsize.c:579 # https://github.com/greenplum-db/gpdb/blob/6.19.0/src/backend/utils/adt/dbsize.c#L579
2&: select pg_total_relation_size('ptest_1_prt_1'::regclass);

# Try to stop txn commit on some segment in the first session up to catalog cache invalidation (https://github.com/greenplum-db/gpdb/blob/6.19.0/src/backend/access/transam/twophase.c#L1528-L1538)
# but after the routine that removes PGPROC entry from ProcArray (https://github.com/greenplum-db/gpdb/blob/6.19.0/src/backend/access/transam/twophase.c#L1486)
# so that renewed snapshots instantiated in concurrent transactions have treated current transaction as completed
(gdb) break twophase.c:1496
1&: commit;

# Issue some dummy DDL query to inspire invalidation of catalog snapshot inside catalog cache invalidation routine (https://github.com/greenplum-db/gpdb/blob/6.19.0/src/backend/utils/cache/inval.c#L596)
# later in the second session
3: create table dummy_test(i int);

# Resume second session
# Cache invalidation occurs within `LockRelationOid()` function on pg_aoseg oid
# `try_relation_open()` call returns NULL value
# SIGSEGV occurs inside `calculate_indexes_size()` function upon attempt to deference incoming relation parameter that is NULL 
(gdb) continue
```

Technically, we might inject the guard condition verifying that returned from `try_relation_open()` relation object is not NULL. But the current case is precedent that there is a danger in early releasing locks on partitions before transaction completion. The similar issue as above segfault might happen anywhere when we execute concurrent locking operation over partitioned table and selection of some partition. Moreover, the motivation of immediate lock releasing for partitions that is exposed in [comment block](https://github.com/greenplum-db/gpdb/blob/6.19.0/src/backend/catalog/dependency.c#L1349-L1352):
```
This helps to keep the lock table size in check, if you e.g. drop a table with thousands of partitions. It would be unpleasant if you could not drop such a table because the lock table runs out of space.
```
doesn't look so ultimate. We might increase `max_locks_per_transaction` GUC to accommodate required number of locks to perform full locking of partitioning hierarchy.
And finally, GPDB master as well as PG master don't optimize anyhow the locking of partitions - all locks are held until transaction completion.

Summing up I propose to remove the logic of immediately releasing locks on partitions from all places but not sure about correctness of this decision and how it will affect other logic.

An comments and thoughts?

**Comment by kainwen** [comment]
Thanks for your reporting!

I have seen several issues that releasing locks on leafs and there is a thread on this https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/-FELESD8BCc/m/EqmOfYrnAgAJ

1. What is the Greenplum version?
2. Recently we found `partition_indexing` case flaky due to almost the same reason

I personally agree with your proposal. Let's wait for more ideas.

**Comment by maksm90** [comment]
Hi, @kainwen ! Thanks for your fast answer.

> I have seen several issues that releasing locks on leafs and there is a thread on this https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/-FELESD8BCc/m/EqmOfYrnAgAJ

Thanks for pointing this link. AFAICS @hlinnaka in PR https://github.com/greenplum-db/gpdb/pull/10468 have already removed early lock releasing logic for all kind of operations required locking (not just for DROP TABLE one as declared in PR title as it seemed to me).

> What is the Greenplum version?

Our case with segfault is reproduced on 6X branch but 5X also have the logic of early lock releasing for partitions and `pg_total_relation_size()` call fails with error (I have been able to reproduce assert failure).

> I personally agree with your proposal. Let's wait for more ideas.

OK, generally I'm ready to backport Heikki's commit https://github.com/greenplum-db/gpdb/pull/10468/commits/86ded366414f10f53c8bfec4e18973c8f33e6c0d to 6X.

**Comment by dpandhi-git** [comment]
PR https://github.com/greenplum-db/gpdb/pull/13036 closed for 6X as won't fixed.
Problem does not exist on Main(7X) branch
